### PR TITLE
Implement 3d texture support for existing functions

### PR DESCRIPTION
**Pull request by ghanzouri**

### CODE REVIEW

**Review by shahingohar — APPROVED** [review]

File: src/ppx/graphics_util.cpp [hunk 1]
```diff
@@ -741,7 +748,7 @@ Result CreateImageFromFile(
             ppxres = CreateImageFromBitmapGpu(pQueue, &bitmap, ppImage, options);
```
Comment (jyn15) [inline src/ppx/graphics_util.cpp:750]: return error if imageDepth != 1 since it's not implemented yet for the Gpu version

Comment (jyn15) [reply src/ppx/graphics_util.cpp:750]: alternatively as we discussed, you can skip changing this function `CreateImageFromFile` so you don't need to worry about the GPU version. Only change `CreateImageFromBitmap` and use that function in our project to create  3D image.

File: src/ppx/graphics_util.cpp [hunk 2]
```diff
@@ -775,6 +782,68 @@ Result CreateImageFromFile(
...
+    ppxres = Bitmap::Create(static_cast<uint32_t>(imageWidth), static_cast<uint32_t>(imageHeight), format, pBitmap);
+    size_t nbytes = Bitmap::StorageFootprint(static_cast<uint32_t>(imageWidth), static_cast<uint32_t>(imageHeight), format); 
+
+    // Fill array data to bitmap at Red channel
+    char* pSrc = (char *)array;
+    char* pDst  = pBitmap->GetData();
+    const uint32_t srcChannelSize = sizeof(float);
+    const uint32_t dstPixelStride = pBitmap->GetPixelStride();
+    for (uint32_t y = 0; y < pBitmap->GetHeight(); ++y) {
+        for (uint32_t x = 0; x < pBitmap->GetWidth(); ++x) {
```
Comment (jyn15) [inline src/ppx/graphics_util.cpp:819]: above line 815, avoid any row stride issue by always re-calculating pDst:

pDst = y * pBitmap->GetRowStride();

File: src/ppx/graphics_util.cpp [hunk 3]
```diff
@@ -775,6 +782,68 @@ Result CreateImageFromFile(
...
+    Result ppxres;
+    // Create bitmap from array
+    Bitmap bitmap;
+    Bitmap* pBitmap = &bitmap;
+    Bitmap::Format format = Bitmap::FORMAT_RGBA_FLOAT;
+    ppxres = Bitmap::Create(static_cast<uint32_t>(imageWidth), static_cast<uint32_t>(imageHeight), format, pBitmap);
+    size_t nbytes = Bitmap::StorageFootprint(static_cast<uint32_t>(imageWidth), static_cast<uint32_t>(imageHeight), format); 
+
+    // Fill array data to bitmap at Red channel
+    char* pSrc = (char *)array;
```
Comment (jyn15) [inline src/ppx/graphics_util.cpp]: reinterpret_cast<char*>

File: src/ppx/graphics_util.cpp [hunk 4]
```diff
@@ -775,6 +782,68 @@ Result CreateImageFromFile(
...
+    const uint32_t srcChannelSize = sizeof(float);
+    const uint32_t dstPixelStride = pBitmap->GetPixelStride();
+    for (uint32_t y = 0; y < pBitmap->GetHeight(); ++y) {
+        for (uint32_t x = 0; x < pBitmap->GetWidth(); ++x) {
+            memcpy(pDst, pSrc, srcChannelSize);
+            pSrc += srcChannelSize;
+            pDst += dstPixelStride;
+        }
+    }
+    free(array);
```
Comment (jyn15) [inline src/ppx/graphics_util.cpp]: I think usually the caller frees the data? Or make it clear in the function comment

File: src/ppx/graphics_util.cpp [hunk 5]
```diff
@@ -775,6 +782,68 @@ Result CreateImageFromFile(
 
 // -------------------------------------------------------------------------------------------------
 
+Result CreateImageFromArray(
```
Comment (jyn15) [inline src/ppx/graphics_util.cpp:789]: Please add some comment about the function, e.g. the source data has single channel and we are writing to the R channel of an RGBA image

**Review by jyn15 — CHANGES_REQUESTED** [review]

File: src/ppx/graphics_util.cpp (hunk 2 above)
Comment (ghanzouri) [reply src/ppx/graphics_util.cpp:819]: I am getting a segmentation fault

File: src/ppx/graphics_util.cpp (hunk 4 above)
Comment (ghanzouri) [reply src/ppx/graphics_util.cpp]: The function Bitmap::LoadFile() from bitmap.cpp creates a pointer float* pData, then free it at row #624  `stbi_image_free(pData);`, should I be doing the same?

**Review by shahingohar — APPROVED** [review]

File: src/ppx/graphics_util.cpp (hunk 4 above)
Comment (shahingohar) [reply src/ppx/graphics_util.cpp]: As Jun mentioned Allocating a memory in one place and deallocating elsewhere is a Big no.

File: src/ppx/graphics_util.cpp [hunk 6]
```diff
@@ -767,7 +768,7 @@ Result CreateImageFromFile(
     else {
         ppxres = Result::ERROR_IMAGE_FILE_LOAD_FAILED;
     }
-
+
```
Comment (jyn15) [inline src/ppx/graphics_util.cpp:771]: avoid spaces/line breaks changes if possible, the original code might have its own style

File: src/ppx/graphics_util.cpp (hunk 2 above)
Comment (jyn15) [reply src/ppx/graphics_util.cpp:819]: not sure could be caused by wrong pDst memory address, but incrementing by dstPixelStride is fine in this case

**Review by jyn15 — APPROVED** [review]

**Pull request merged**